crates/admin/src/cluster_controller/observed_cluster_state.rs

```diff
@@ -8,86 +8,84 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::collections::{HashMap, HashSet};
-
-use restate_types::cluster::cluster_state::{ClusterState, NodeState, RunMode};
+use restate_types::cluster::cluster_state::{ClusterState, RunMode};
 use restate_types::identifiers::PartitionId;
 use restate_types::{GenerationalNodeId, NodeId, PlainNodeId};
+use std::collections::{HashMap, HashSet};
 
 /// Represents the scheduler's observed state of the cluster. The scheduler will use this
 /// information and the target scheduling plan to instruct nodes to start/stop partition processors.
 #[derive(Debug, Default, Clone)]
 pub struct ObservedClusterState {
-    pub partitions: HashMap<PartitionId, ObservedPartitionState>,
-    pub alive_nodes: HashMap<PlainNodeId, GenerationalNodeId>,
-    pub dead_nodes: HashSet<PlainNodeId>,
-    pub nodes_to_partitions: HashMap<PlainNodeId, HashSet<PartitionId>>,
+    partitions: HashMap<PartitionId, ObservedPartitionState>,
+    // The logic that uses ObservedClusterState assumes that it has a snapshot of the current alive
+    // nodes. We could use restate_core::cluster_state::ClusterState directly, if we kept the read
+    // lock. However, since ObservedClusterState is hold across await points, this would prevent
+    // any concurrent updates to the ClusterState. Therefore, we keep a separate copy of the alive
+    // nodes here until the users of ObservedClusterState no longer require a snapshot.
+    alive_nodes: HashMap<PlainNodeId, GenerationalNodeId>,
+    nodes_to_partitions: HashMap<PlainNodeId, HashSet<PartitionId>>,
 }
 
 impl ObservedClusterState {
-    pub fn is_node_alive(&self, node_id: impl Into<NodeId>) -> bool {
+    pub fn partition_state(&self, partition_id: &PartitionId) -> Option<&ObservedPartitionState> {
+        self.partitions.get(partition_id)
+    }
+
+    pub fn alive_nodes(&self) -> impl Iterator<Item = &GenerationalNodeId> {
+        self.alive_nodes.values()
+    }
+
+    pub fn alive_nodes_empty(&self) -> bool {
+        self.alive_nodes.is_empty()
+    }
+
+    /// Returns the generational node id if the given node_id is alive.
+    pub fn is_node_alive(&self, node_id: impl Into<NodeId>) -> Option<GenerationalNodeId> {
         let node_id = node_id.into();
 
         match node_id {
-            NodeId::Plain(plain_node_id) => self.alive_nodes.contains_key(&plain_node_id),
+            NodeId::Plain(plain_node_id) => self.alive_nodes.get(&plain_node_id).cloned(),
             NodeId::Generational(generational_node_id) => self
                 .alive_nodes
                 .get(&generational_node_id.as_plain())
-                .is_some_and(|node_id| *node_id == generational_node_id),
+                .filter(|node_id| **node_id == generational_node_id)
+                .cloned(),
```

Comment:
nit: would `.copied()` work here?

Reply:
Yes, I think it should. Thanks for making me aware of it. Will update it :-)